change config keys, support timestamp and json

potato2003 opened the pull request.

potato2003 committed:
change config keys according to de-fact standard, other plugins

potato2003 committed:
support timestamp, json format

potato2003 committed:
update README

sonots commented on README.md:
```diff
@@ -13,10 +13,11 @@
 - worksheet_gid (integer, default: 0): worksheet's gid if you want to specific worksheet
 - mode (string, default: append): writing record method, available mode are `append` and `replace`
-- is_write_header (bool, default: false): if true, write header to first record
+- is_write_header_line (bool, default: false): if true, write header to first record
```
http://www.embulk.org/docs/built-in.html#csv-formatter-plugin uses `header_line`. Having the same name with such standard plugins is better.

sonots commented on embulk-output-google_spreadsheets.gemspec:
```diff
@@ -18,4 +18,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'bundler', ['>= 1.10.6']
   spec.add_development_dependency 'rake', ['>= 10.0']
   spec.add_dependency 'google_drive'
+  spec.add_dependency 'activesupport'
```
Depending on activesupport is always bad idea.

sonots commented on README.md:
```diff
@@ -13,10 +13,11 @@
   notation.
-- null_representation (string, default: ''): replace null to `null_representation`
+- null_string (string, default: ''): replace null to `null_string`
+- default_timezone (string, default: 'UTC'): time zone of timestamp columns
```
I want `default_timestamp_format` too so that we can specify the format.

sonots replied:
What we need to support is timezone strings like `+00:00`, `+09:00`.
Supporting `UTC` specially is good idea, but we do not need to support something like `CST` because we can not recognize whether they are China Standard Time or (America) Central Standard Time.

We can use `Time#localtime`.
Note that it destructively changes the timezone of the object (although it does not have !), So we write

```
time.dup.localtime("+09:00").strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
```

in the case of filter plugin, but this is an output plugin (the last plugin), it should be okay to be destructive the time object, thus,

```
time.localtime("+09:00").strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
```

sonots commented on lib/embulk/output/google_spreadsheets.rb:
```diff
@@ -142,8 +147,16 @@ def self.clean_previous_records(worksheet, schema, task)
+
+        case type
+        when :timestamp
+          v.in_time_zone.strftime('%Y-%m-%d %H:%M:%S %Z')
```
we, embulk people, use `%Y-%m-%d %H:%M:%S.%6N %z` as default format.

sonots commented on lib/embulk/output/google_spreadsheets.rb:
```diff
@@ -50,9 +53,11 @@ def init
 
         @worksheet = self.class.build_worksheet_client(task)
+
+        Time.zone = task["default_timezone"]
```
This is very very bad idea. `Time` is shared among all plugins and embulk itself.

sonots commented:
Can you change parameters like https://github.com/medjed/embulk-input-google_spreadsheets?
`spreadsheets_url`, `worksheet_title`, `start_column`, `start_row`, and supports `auth_method`.

sonots commented:
Additional request: Can you support a way to write json_keyfile content on yaml directly as:

https://github.com/medjed/embulk-input-google_spreadsheets/blob/master/lib/embulk/input/google_spreadsheets.rb#L13-L24
https://github.com/medjed/embulk-input-google_spreadsheets/blob/master/lib/embulk/input/google_spreadsheets.rb#L104

potato2003 committed:
fix timezone

- remove activesupport
- Time.zone should not be changed by plugin, since it will affect an outside of plugin

potato2003 committed:
add options according to embulk-input-google_spreadsheets

sonots commented on README.md:
```diff
@@ -9,14 +9,17 @@
+- start_column (integer, default: 1): specific the start column
+- null_string (string, default: ''): replace null to `null_string`
+- default_timezone (string, default: '+00:00'): time zone offset of timestamp columns
+- default_timezone_format (string, default: '%Y-%m-%d %H:%M:%S.%6N %z'): output format of timestamp columns
```
timestamp_format

sonots commented on lib/embulk/output/google_spreadsheets.rb:
```diff
@@ -134,16 +166,27 @@ def self.clean_previous_records(worksheet, schema, task)
+          zone_offset = task['default_timezone']
+          format      = task['default_timezone_format']
+
+          v.dup.localtime(zone_offset).strftime(format)
```
localtime("UTC") fails. We need a special care for "UTC", just skipping localtime()

potato2003 committed:
fix timezone_format to timestamp_format

potato2003 committed:
allow 'UTC' of default_timezone config

sonots commented on lib/embulk/output/google_spreadsheets.rb:
```diff
@@ -134,16 +166,29 @@ def self.clean_previous_records(worksheet, schema, task)
+        case type
+        when :timestamp
+          zone_offset = task['default_timezone']
+          zone_offset = '+00:00' if zone_offset == 'UTC'
```
We should do such conversion only one time I this plugin rather than doing for each value.

How about making an instance method which memoizes the converted value with an instance variable.

potato2003 committed:
fix default value

potato2003 committed:
avoid checking per record

sonots commented:
LGTM!!
